processing/src/main/java/org/apache/druid/segment/join/PossiblyNullDimensionSelector.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment.join;
+
+import com.google.common.base.Predicate;
+import org.apache.druid.query.filter.ValueMatcher;
+import org.apache.druid.query.monomorphicprocessing.RuntimeShapeInspector;
+import org.apache.druid.segment.AbstractDimensionSelector;
+import org.apache.druid.segment.DimensionSelector;
+import org.apache.druid.segment.DimensionSelectorUtils;
+import org.apache.druid.segment.IdLookup;
+import org.apache.druid.segment.data.IndexedInts;
+import org.apache.druid.segment.data.ZeroIndexedInts;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.function.BooleanSupplier;
+
+public class PossiblyNullDimensionSelector extends AbstractDimensionSelector implements IdLookup
+{
+  private final DimensionSelector baseSelector;
+  private final BooleanSupplier beNull;
+  private final NullAdjustedIndexedInts nullAdjustedRow;
+
+  // 1 for selectors that have dictionaries (cardinality >= 0), but don't contain null.
+  private final int nullAdjustment;
+
+  public PossiblyNullDimensionSelector(DimensionSelector baseSelector, BooleanSupplier beNull)
+  {
+    this.baseSelector = baseSelector;
+    this.beNull = beNull;
+
+    if (baseSelector.nameLookupPossibleInAdvance() &&
+        (baseSelector.getValueCardinality() == 0 || (baseSelector.getValueCardinality() > 0
+                                                     && baseSelector.lookupName(0) != null))) {
+      this.nullAdjustment = 1;
+    } else {
+      this.nullAdjustment = 0;
+    }
+
+    this.nullAdjustedRow = new NullAdjustedIndexedInts(nullAdjustment);
+  }
```

Comment:
Some comments would be helpful to understand + unit tests for maintainability

Reply:
I've added a bunch of tests.